### DISCUSSION TITLE
Early feedback - Astro.request

### DISCUSSION THREAD

**Discussion by matthewp** [withastro/roadmap#151] [Proposal]
I would like early feedback on a problem I'm not sure what the best way to resolve is.

In SSR you want access to headers in order to check for a user's cookie to see if they are logged in, etc.

My first thought was to expose a [Request](https://developer.mozilla.org/en-US/docs/Web/API/Request) object as `Astro.request`.

However we already have an `Astro.request` that is shaped like this:

```ts
interface AstroRequest {
  /** get the current page URL */
  url: URL;

  /** get the current canonical URL */
  canonicalURL: URL;

  /** get page params (dynamic pages only) */
  params: Params;
}
```

My first thought was to make AstroRequest extend Request and add `canonicalURL` and `params`. However on Request, `url` is a string, not a URL. So we can't make this a non-breaking change.

## Options

1. __breaking__ - Make AstroRequest extend Request, but `url` is now a string.
2. __breaking__ - Make `Astro.request` be a plain Request, no extending, and move `canonicalURL` and `params` somewhere else.
  - If this one, where should those live?
4. __non-breaking__ - Put the Request instance somewhere else. Where?
  - `Astro.request.raw`
  - `Astro.webRequest`
5. __non-breaking__ - Do not expose a Request anywhere, instead add `headers` to `Astro.request` and allow it to continue to be a plain object.
  - Simplest solution probably, but doesn't align well with having a `Response` that can be used.
  - We also need `method`, not just headers.

**Comment by natemoo-re** [comment]
> 2. breaking - Make Astro.request be a plain Request, no extending, and move canonicalURL and params somewhere else.

I'd be most in favor of this option personally!

**Reply by FredKSchott** [reply]
Yea, I think I like this as well. I think `params` could live elsewhere (ex: `Astro.params`, aka the Next.js approach: https://nextjs.org/docs/api-reference/data-fetching/get-server-side-props#context-parameter) 

Or, we go the Remix route and document how you can parse the params and create canonical-URLs yourself: https://remix.run/docs/en/v1/api/conventions#loader-request

```
  // read a cookie
  const cookie = request.headers.get("Cookie");

  // parse the search params
  const url = new URL(request.url);
  const search = url.searchParams.get("search");
```

**Reply by natemoo-re** [reply]
IIRC `params` in our case is not `URLSearchParams` but file-system route-based `params`.

**Reply by matthewp** [reply]
Yeah, but Astro.params seems fine to me. The only place you use those are in dynamic routes.

canonicalURL could be created yourself, but this is what it would take:

```astro
---
const url = new URL(Astro.request.url);
const canonicalURL = new URL(url.pathname, Astro.site);
---
```

**Comment by matthewp** [comment]
This is now an RFC: https://github.com/withastro/rfcs/pull/153

**Comment by matthewp** [comment]
This is an old issue that is no longer relevant.
